src/liboslcomp/typecheck.cpp

```diff
@@ -1180,58 +1156,422 @@ ASTfunction_call::typecheck_struct_constructor ()
 }
 
 
+/// <doc CandidateFunctions>
+///
+/// Score a set of polymorphic functions based on arguments & return type.
+///
+/// When choosing which varaint to use precedence is given to arguments and the
+/// return type is used as a possible tie breaker in case of any ambiguity.
+/// The highest score for an argument is given when it is an exact match.
+/// If the argument needs to be coerced into a type OSL will prefer going
+/// from any triple-types to any other triple-type before promotion of a
+/// float to a triple.  When converting one triple to another triple, OSL will
+/// prefer to coerce from one of the spatial triples (vector/point/normal) to
+/// another spatial triple before coercion from a spatial triple to color.
+///
+/// Float to int coercion is scored, but is currently a synmonym for kNoMatch
+/// as the spec does not allow implicit float to int conversion.
+///
+/// </script>
+/// ~~~~
+class CandidateFunctions {
+    enum {
+        kExactMatch     = 100,
```

Comment:
So higher scores are closer matches? Do you think that for future proofing (imagine a future with default arguments or ... style arguments) it might be better to have it the other way around? In other words, if 0 was perfect match, and each substitution had a "cost", summing the cost over the arguments? Because as you have it now (if I understand properly), you are summing score over the argument and aiming for high score, so "more arguments" naturally seems to score higher, whereas if perfect match is 0, then you are more directly optimizing (fewer errors) if you know what I mean.

I don't have a strong feeling about this one way or the other, but just pointing it out in case we want to reverse the sense of it now.

Reply:
More arguments does lead to a higher score, but having too few or too many will force the score for that variant to 0.

I think scoring needs to go up to support variadic args (or at least it's unclear to me how to do the following in your model)
```
A. void printer(string fmt, ...);
B. void printer(string prefix, string fmt, ...);

printer("FORMAT", 2) // should call A
printer("PREFIX", "FORMAT") // should call B
```